server/lib/device/device.getFeatureStates.js

```diff
@@ -0,0 +1,117 @@
+ */
+async function getFeatureStates(options) {
+  const optionsWithDefault = Object.assign({}, DEFAULT_OPTIONS, options);
+
+  // Build query for not feature_state info
+  const queryParams = {
+    include: [
+      {
+        model: db.DeviceFeature,
+        as: 'features',
+      },
+      {
+        model: db.Room,
+        as: 'room',
+      },
+    ],
+    offset: optionsWithDefault.skip,
+  };
+
+  // Fix attributes
+  if (optionsWithDefault.attributes_device && optionsWithDefault.attributes_device.length > 0) {
+    queryParams.attributes = optionsWithDefault.attributes_device;
+  }
+  if (optionsWithDefault.attributes_device_feature && optionsWithDefault.attributes_device_feature.length > 0) {
+    queryParams.include[0].attributes = optionsWithDefault.attributes_device_feature;
+  }
+  if (optionsWithDefault.attributes_device_room && optionsWithDefault.attributes_device_room.length > 0) {
+    queryParams.include[1].attributes = optionsWithDefault.attributes_device_room;
+  }
+
+  // Search by feature selector
+  if (optionsWithDefault.device_feature_selector) {
+    if (optionsWithDefault.device_feature_selector.length > 1) {
+      queryParams.include[0].where = {
+        selector: { [Op.in]: optionsWithDefault.device_feature_selector },
+      };
+    } else {
+      queryParams.include[0].where = {
+        selector: { [Op.eq]: optionsWithDefault.device_feature_selector },
+      };
+    }
+  }
+
+  const devices = await db.Device.findAll(queryParams);
+  const devicesPlain = await devices.map((device) => device.get({ plain: true }));
+
+  if (devicesPlain && devicesPlain.length > 0) {
+    for (let i = 0; i < devicesPlain.length; i += 1) {
+      const device = devicesPlain[i];
+      if (device.features && device.features.length > 0) {
+        for (let j = 0; j < device.features.length; j += 1) {
+          const feature = device.features[j];
+          // Search feature state of current feature
+          let beginStateSearchDate = optionsWithDefault.begin_date;
+          let currentState = [];
+
+          // If column last_downsampling is explicitly asked
+          // => part of data is extract from device_feature_state_light
+          if (
+            feature.last_downsampling &&
+            optionsWithDefault.attributes_device_feature &&
+            optionsWithDefault.attributes_device_feature.length > 0 &&
+            optionsWithDefault.attributes_device_feature.includes('last_downsampling')
+          ) {
+            const queryStateLightParams = {
+              where: {
+                device_feature_id: { [Op.eq]: feature.id },
+                created_at: { [Op.between]: [optionsWithDefault.begin_date, feature.last_downsampling] },
+              },
+              order: [['created_at', 'ASC']],
+              raw: true,
+            };
+            // eslint-disable-next-line no-await-in-loop
+            currentState = currentState.concat(await db.DeviceFeatureStateLight.findAll(queryStateLightParams));
```

Comment:
I do'nt use r statement because I consider maximum f 2 (maybe 3) features reasonably used in multi feature chart because size of box is very small (more of 3 feature will be unreadable in my opinion) and with 1 request by features I have already feature state result ready to insert in feature object without post treatment 

I can try with one request

Reply:
I change it to put one request with or statement